src/main/src/main/java/org/geoserver/catalog/GeoServerSLDVisitor.java

```diff
@@ -0,0 +1,469 @@
+import org.geotools.factory.Hints;
+import org.geotools.feature.SchemaException;
+import org.geotools.referencing.crs.DefaultGeographicCRS;
+import org.geotools.styling.AbstractStyleVisitor;
+import org.geotools.styling.FeatureTypeConstraint;
+import org.geotools.styling.NamedLayer;
+import org.geotools.styling.NamedStyle;
+import org.geotools.styling.RemoteOWS;
+import org.geotools.styling.Style;
+import org.geotools.styling.StyledLayer;
+import org.geotools.styling.StyledLayerDescriptor;
+import org.geotools.styling.UserLayer;
+import org.opengis.feature.Feature;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+import org.opengis.feature.type.FeatureType;
+import org.opengis.feature.type.Name;
+import org.opengis.filter.Filter;
+import org.opengis.referencing.FactoryException;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+import org.opengis.referencing.operation.TransformException;
+import org.opengis.util.ProgressListener;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Visitor for standalone {@link org.geotools.styling.StyledLayerDescriptor}s
+ *
+ * Visits each {@link StyledLayer}s of an SLD.
+ * Resolves LayerGroups, Remote OWS references, and inline features into individual layers
+ * Visits each {@link Style} in each{@link StyledLayer}, alongside each of these resolved layers
+ *
+ * Intended to provide a definitive, extensible approach to parsing standalone
+ * {@link org.geotools.styling.StyledLayerDescriptor}s, including style groups and external SLDs.
+ */
+public abstract class GeoServerSLDVisitor extends AbstractStyleVisitor {
+
+    protected static Logger LOGGER = org.geotools.util.logging.Logging.getLogger("org.geoserver.catalog");
+
+    protected final Catalog catalog;
+    protected final CoordinateReferenceSystem fallbackCrs;
+
+    //Current state
+    protected PublishedInfo info;
+    protected StyledLayer layer;
+    protected int styleCount = 0;
+
+
+    public GeoServerSLDVisitor(Catalog catalog, CoordinateReferenceSystem fallbackCrs) {
+        this.catalog = catalog;
+        this.fallbackCrs = fallbackCrs;
+    }
+
+    /**
+     * Called on each named layer in the style
+     *
+     * Implementations should resolve and return the corresponding PublishedInfo
+     *
+     * @param namedLayer The named layer
+     * @return The resolved catalog layer
+     */
+    public abstract PublishedInfo visitNamedLayerInternal(StyledLayer namedLayer);
+
+    /**
+     * Called on each remote OWS user layer in the style
+     *
+     * @param userLayer The user layer
+     */
+    public abstract void visitUserLayerRemoteOWS(UserLayer userLayer);
```

Comment:
Are you in position to provide a default method? Or do you want to force subclasses to think about it.

Reply:
I've got an adapter subclass; I suppose I could have done that using default methods instead.